Assignment1Q1.java

```diff
@@ -0,0 +1,110 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+public class Assignment1Q1 {
+
+    private static final int CHARS_AMOUNT = 26;
+
+    /**
+     * Performs basic validation.
+     */
+    private static boolean validate(String str1, String str2) {
+        if (str1 == null || str2 == null) { // null strings are not permutation of each other
+            return false;
+        }
+        if (str1.length() != str2.length()) { // strings with different length are not permutations of each other
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * More general solution that works for any kind of characters.
+     */
+    private static boolean isPermutation(String str1, String str2) {
+        if (!validate(str1, str2)) {
+            return false;
+        }
+        if (str1.isEmpty() && str2.isEmpty()) { // simple case
+            return true;                        // empty strings are permutation of each other
+        }
+
+        Function<String, Map<Character, Integer>> fillCounter = (String str) -> {

```

Comment:
while Java8 functions are nice, I would have just added this as a static method. Having it as a Function complicates things and doesn't gain you anything that I can see.